### PR TITLE
gh-119609: Add PyUnicode_Export() and PyUnicode_Import() functions

### PR DESCRIPTION
**Pull request by vstinner**
* Issue: gh-119609

----
📚 Documentation preview 📚: https://cpython-previews--119610.org.readthedocs.build/

### CODE REVIEW

**Comment by vstinner** [comment]
Docs build failed because of https://github.com/python/cpython/pull/119607

**Comment by vstinner** [comment]
> Docs build failed because of https://github.com/python/cpython/pull/119607

PR rebased on top of this fix.

File: Objects/unicodeobject.c [hunk 1]
```diff
@@ -2094,6 +2094,89 @@ _PyUnicode_FromUCS4(const Py_UCS4 *u, Py_ssize_t size)
...
+}
+
+PyObject*
+PyUnicode_FromNativeFormat(const void *data, Py_ssize_t size,
+                           int native_format)
+{
+    if (size < 0) {
+        PyErr_SetString(PyExc_ValueError, "Negative size");
+        return NULL;
+    }
```
Comment (erlend-aasland) [inline Objects/unicodeobject.c]: FTR, this check is already handled by `PyUnicode_New`.

**Comment by zooba** [comment]
Petr indicated on the issue that he's got Thoughts and will write them up this week.

File: Objects/unicodeobject.c (hunk 1 above)
Comment (vstinner) [reply Objects/unicodeobject.c]: `PyUnicode_New()` raises SystemError. I would like to raise ValueError here.

**Comment by vstinner** [comment]
@encukou: Please review the updated PR.

@encukou: I prefer to make sure that the exported string ends with a NUL character, rather than making sure that it's not the case. It's convenient and cheap.

I rebased the PR on the main branch, I fixed merge conflicts, I updated the doc for new function names, I included some of Petr's suggestions. I marked the PR as ready for review (it's no longer a draft).

File: Doc/c-api/unicode.rst [hunk 2]
```diff
@@ -341,6 +341,63 @@ APIs:
...
+   The export must be released by :c:func:`PyUnicode_ReleaseExport`.
+
+   *unicode*, *nbytes* and *format* must not be NULL.
+
+   Available formats:
+
+   .. c:namespace:: NULL
+
+   ===================================  ========  ===========================
+   Constant Identifier                  Value  Description
```
Comment (encukou) [inline Doc/c-api/unicode.rst]:
```suggestion
   Constant Identifier                    Value  Description
```

File: Doc/c-api/unicode.rst [hunk 3]
```diff
@@ -341,6 +341,63 @@ APIs:
...
+
+   Export the contents of the *unicode* string in one of the requested format
+   *requested_formats*.
+
+   * On success, set *\*nbytes* and *\*format*, and return the contents.
+   * On error, set an exception and return ``NULL``.
+
+   The contents is valid as long as *unicode* is valid.
+
+   The export must be released by :c:func:`PyUnicode_ReleaseExport`.
```
Comment (encukou) [inline Doc/c-api/unicode.rst:354]:
```suggestion
   The export must be released by :c:func:`PyUnicode_ReleaseExport`.
   The contents of the buffer are valid until they are released.

   The buffer must not be modified.

   To help avoid buffer overflows, the exported buffer is followed
   by a zero codepoint. This terminator is not counted in *nbytes*. 
```

File: Doc/c-api/unicode.rst [hunk 4]
```diff
@@ -341,6 +341,63 @@ APIs:
...
+   in the table above.
+
+   Note that future versions of Python may introduce additional formats.
+
+   .. versionadded:: 3.14
+
+
+.. c:function:: void PyUnicode_ReleaseExport(PyObject *unicode, const void* data, uint32_t format)
+
+   Release an export created by :c:func:`PyUnicode_Export`.
```
Comment (encukou) [inline Doc/c-api/unicode.rst:386]:
```suggestion
   Release an export created by :c:func:`PyUnicode_Export`.

   Each argument must match the corresponding argument or result of
   a single earlier call to :c:func:`PyUnicode_Export`.
   In particular, this means that you must hold a reference to *unicode*
   while an export is valid.
```

File: Doc/c-api/unicode.rst [hunk 5]
```diff
@@ -341,6 +341,63 @@ APIs:
...
+   ===================================  ========  ===========================
+   .. c:macro:: PyUnicode_FORMAT_ASCII  ``0x01``  ASCII string (``Py_UCS1*``)
+   .. c:macro:: PyUnicode_FORMAT_UCS1   ``0x02``  UCS-1 string (``Py_UCS1*``)
+   .. c:macro:: PyUnicode_FORMAT_UCS2   ``0x04``  UCS-2 string (``Py_UCS2*``)
+   .. c:macro:: PyUnicode_FORMAT_UCS4   ``0x08``  UCS-4 string (``Py_UCS4*``)
+   .. c:macro:: PyUnicode_FORMAT_UTF8   ``0x10``  UTF-8 string (``char*``)
+   ===================================  ========  ===========================
+
+   *requested_formats* can be a single format or a combination of the formats
+   in the table above.
```
Comment (encukou) [inline Doc/c-api/unicode.rst]:
```suggestion
   *requested_formats* can be a single format or a bitwise combination of the
   formats in the table above.
   On success, *\*format* will be set to a single one of the requested flags.
```

**Review by encukou — COMMENTED** [review]
Looks good to me!
(I don't know about the other C API WG members.)

**Comment by vstinner** [comment]
@davidism: Does the proposed API solve your MarkupSafe use case?

**Comment by vstinner** [comment]
I created [Add PyUnicode_Export() and PyUnicode_Import() to the limited C API](https://github.com/capi-workgroup/decisions/issues/33) issue in the C API WG Decisions project.

**Comment by davidism** [comment]
I really appreciate you thinking about this after talking at PyCon! From what I can tell (I'm not very familiar with C), the MarkupSafe code would mostly remain the same, but use the two new functions instead of the existing non-abi3 functions? Perhaps you could show a brief example or a high level explanation of what I would change? https://github.com/pallets/markupsafe/blob/d12057361ad75c4569e2e61712c234acc69d5d0b/src/markupsafe/_speedups.c

**Comment by vstinner** [comment]
@davidism:
> I really appreciate you thinking about this after talking at PyCon!

You're welcome.

> From what I can tell (I'm not very familiar with C), the MarkupSafe code would mostly remain the same, but use the two new abi3 functions instead of the existing functions? Perhaps you could show a brief example or a high level explanation of what I would change?

I created a PR to show how these functions can be used: https://github.com/pallets/markupsafe/pull/451

The stable ABI is less efficient since it requires to allocate a UCS1/UCS2/UCS4 buffer first, write into the buffer, and only then create a Python str object from this buffer. That's because PyUnicode_New() is excluded from the stable ABI.

**Comment by vstinner** [comment]
I wrote a different PR using `Py_buffer`: https://github.com/python/cpython/pull/123738. I close this PR for now.

**Pull request closed without merging**